File: P2/Plan.cpp
/**
 @file Plan.cpp
 @author Jakob Balkovec (CPSC 3200)
 @instructor A. Dingle (CPSC 3200)

 @date Wed 20th Sep

 @version
 Revision History
 - 1.0 [07/10/2023]: Class Methods, Move semantics
 - 2.0 [08/10/2023]: Debugging, Refine documentatin
 - 3.0 [09/10/2023]: Debugging

 @brief
 * This file contains the implementation of the Plan class, which represents
 * handling an array of Formula Objects. The plan class
 * encapsulates information about the array and abstratcs memory managment.

 @invariant
 * Resource Arrays should contain valid Strings and should not be empty
 * Quantity Arrays should contain Non-Negative Quantities and should not be empty
 * Proficiency Level should be Non-Negative and within valid range
*/
#include <stdexcept>

#include "Formula.h"
#include "Plan.h"

/**
 * @namespace ResourceConversion
 * @brief Defines the Formula and Plan class, avoiding global namespace polution
 */
namespace ResourceConversion
{
  /**
   * @brief Resize the Plan to accommodate a new capacity.
   *
   * @param NewCapacity The new capacity for the Plan.
   *
   * @pre NewCapacity should be greater than or equal to the current Size.
   *
   * @post The Plan's capacity will be updated to NewCapacity.
   *       The Plan's Size remains unchanged, and the existing data is preserved.
   *       FormulaArray points to a new array of size NewCapacity containing the same data.
   */
  void Plan::ResizePlan (size_t NewCapacity)
  {
    Formula* NewFormulaArray = new Formula[NewCapacity];
    for (size_t i = 0; i < Size; ++i)
    {
      NewFormulaArray[i] = FormulaArray[i];
    }
    delete[] FormulaArray;
    FormulaArray = NewFormulaArray;
    Size = Capacity;
    Capacity = NewCapacity;
  }

  /**
   * @brief Clears the Plan by releasing allocated resources and resetting its size and capacity to 0.
   *
   * @pre None.
   *
   * @post All allocated memory for FormulaArray is freed, and FormulaArray is set to nullptr.
   *       The Size and Capacity of the Plan are both set to 0.
   */
  inline void Plan::ClearPlan ()
  {
    if (FormulaArray != nullptr)
    {
      delete[] FormulaArray;
    }
    FormulaArray = nullptr;
    Size = Capacity = 0;
  }

  /**
   * @brief Copies data from another Plan into this Plan.
   *
   * @param other The Plan object from which data will be copied.
   *
   * @pre The 'other' Plan object should be properly initialized.
   *
   * @post This Plan will contain a deep copy of the data from the 'other' Plan,
   *       including the same Size and Capacity.
   */
  inline void Plan::CopyPlanData (const Plan& other)
  {
    Size = other.Size;
    Capacity = other.Capacity;
    FormulaArray = new Formula[Capacity];
    for (size_t i = 0; i < Size; ++i)
    {
      FormulaArray[i] = other.FormulaArray[i];
    }
  }

  /**
   * @brief Resets this Plan by transferring ownership of data from another Plan.
   *
   * @param other The Plan object from which ownership of data will be transferred.
   *
   * @pre The 'other' Plan object should be properly initialized.
   *
   * @post This Plan takes ownership of the data (including Size and Capacity) from 'other.'
   *       'other' becomes an empty Plan with no data, and its Capacity is set to 0.
   */
  inline void Plan::ResetPlan (Plan& other)
  {
    Capacity = other.Capacity;
    Size = other.Size;
    FormulaArray = other.FormulaArray;
    other.Capacity = 0;
    other.Size = 0;
    other.FormulaArray = nullptr;
  }

  /**
   * @brief Constructs an empty Plan with an initial Capacity of 2.
   *
   * @pre None.
   *
   * @post An empty Plan object is created with Capacity set to 2.
   *       The Size is initialized to 0, and an array of Capacity elements is allocated.
   */
  Plan::Plan ()
  {
    Capacity = 2;
    Size = 0;
    FormulaArray = new Formula[Capacity];
  }

  /**
   * @brief Constructs a Plan object with an initial sequence of Formulas and a specified initial size.
   *
   * @param InitialSequence An array of Formula objects representing the initial sequence.
   * @param InitialSize The initial size of the Plan, which must be equal to or greater than the size of InitialSequence.
   *
   * @pre InitialSize should be greater than or equal to the size of InitialSequence.
   *
   * @post A Plan object is constructed with Capacity set to InitialSize.
   *       The Size is initialized to InitialSize, and the data is copied from InitialSequence.
   */
  Plan::Plan (Formula* InitialSequence, size_t InitialSize)
  {
    Capacity = InitialSize;
    Size = InitialSize;
    FormulaArray = new Formula[Capacity];
    for (size_t i = 0; i < Size; ++i)
    {
      FormulaArray[i] = InitialSequence[i];
    }
  }

  /**
   * @brief Constructs a Plan object with an initial sequence of Formulas using an initializer list.
   *
   * @param InitialSequence An initializer list of Formula objects representing the initial sequence.
   *
   * @pre None.
   *
   * @post A Plan object is constructed with Capacity set to the size of InitialSequence.
   *       The Size is initialized to the same size as Capacity, and the data is copied from InitialSequence.
   */
  Plan::Plan (const std::initializer_list<Formula>& InitialSequence)
  {
    Capacity = InitialSequence.size ();
    Size = Capacity;
    FormulaArray = new Formula[Capacity];

    size_t i = 0;
    for (const Formula& FormulaObj : InitialSequence) {
      FormulaArray[i++] = FormulaObj;
    }
  }

  /**
   * @brief Destructor for the Plan class.
   *
   * @pre None.
   *
   * @post The Plan object is destroyed, and any allocated resources are released.
   */
  Plan::~Plan ()
  {
    ClearPlan ();
  }

  /**
   * @brief Copy constructor for the Plan class.
   *
   * @param other The Plan object to be copied.
   *
   * @pre None.
   *
   * @post A new Plan object is created as a deep copy of 'other,' including its data (Size and Capacity).
   */
  Plan::Plan (const Plan& other)
  {
    CopyPlanData (other);
  }

  /**
   * @brief Copy assignment operator for the Plan class.
   *
   * @param other The Plan object to be assigned from.
   *
   * @pre None.
   *
   * @post The current Plan object is assigned the data (including Size and Capacity) from 'other.'
   *       Any existing data in the current object is cleared.
   *
   * @return A reference to the current Plan object after the assignment.
   */
  Plan& Plan::operator=(const Plan& other)
  {
    if (this == &other) { return *this; }

    ClearPlan ();
    CopyPlanData (other);

    return *this;
  }

  /**
   * @brief Move constructor for the Plan class.
   *
   * @param other The Plan object to be moved from.
   *
   * @pre None.
   *
   * @post A new Plan object is created by taking ownership of the data (including Size and Capacity) from 'other.'
   *       'other' becomes an empty Plan with no data, and its Capacity is set to 0.
   */
  Plan::Plan (Plan&& other) noexcept
  {
    std::swap(FormulaArray, other.FormulaArray);
    std::swap(Size, other.Size);
    std::swap(Capacity, other.Capacity);
    
    ResetPlan (other);
  }

  /**
   * @brief Move assignment operator for the Plan class.
   *
   * @param other The Plan object to be assigned from.
   *
   * @pre None.
   *
   * @post The current Plan object is assigned the data (including Size and Capacity) from 'other' via a move operation.
   *       'other' becomes an empty Plan with no data, and its Capacity is set to 0.
   *
   * @return A reference to the current Plan object after the move assignment.
   */
  Plan& Plan::operator=(Plan&& other)
  {
    if (this == &other) { return *this; }

    ClearPlan ();
    CopyPlanData (other);
    ResetPlan (other);

    return *this;
  }

  /**
   * @brief Access the Formula at the specified index.
   *
   * @param Index The index of the Formula to access.
   *
   * @pre Index should be within the valid range [0, Size - 1].
   *
   * @post None.
   *
   * @throw std::out_of_range if Index is out of range.
   *
   * @return A reference to the Formula at the specified index.
   */
  Formula& Plan::operator[](size_t Index)
  {
    if (Index >= Size) { throw std::out_of_range ("[Index out of range]"); }
    return FormulaArray[Index];
  }

  /**
   * @brief Add a new Formula to the Plan.
   *
   * @param NewFormula The Formula to be added.
   *
   * @pre None.
   *
   * @post If adding the Formula exceeds the Capacity, the Plan is resized to accommodate it.
   *       The new Formula is added to the end of the Plan, and the Size is updated accordingly.
   */
  void Plan::AddFormula (const Formula NewFormula) {
    if (Size >= Capacity) {
      ResizePlan (Capacity * 2);
    }
    FormulaArray[Size++] = NewFormula;
  }

  /**
   * @brief Remove the last Formula from the Plan.
   *
   * @pre The Size of the Plan should be greater than 0.
   *
   * @post The last Formula is removed from the Plan, and the Size is decremented.
   *
   * @throw std::invalid_argument if the Size is less than or equal to 0.
   */
  void Plan::RemoveLastFormula ()
  {
    if (Size <= 0) { throw std::invalid_argument ("[Plan size is less than or equal to 0]"); }
    --Size;
  }

  /**
   * @brief Replace the Formula at the specified index with a new Formula.
   *
   * @param Index The index of the Formula to replace.
   * @param NewFormula The new Formula to be assigned at the specified index.
   *
   * @pre Index should be within the valid range [0, Size - 1].
   *
   * @post The Formula at the specified index is replaced with the new Formula.
   *
   * @throw std::out_of_range if Index is out of range.
   */
  void Plan::ReplaceFormula (size_t Index, const Formula& NewFormula)
  {
    if (Index > Size) { throw std::out_of_range ("[Index is Out of Bounds"); }
    FormulaArray[Index] = NewFormula;
  }

  /**
   * @brief Apply all Formulas in the Plan.
   *
   * @pre The Size of the Plan should be greater than 0.
   *
   * @post All Formulas in the Plan are applied sequentially.
   *
   * @throw std::logic_error if the Size is less than or equal to 0.
   */
  void Plan::PlanApply ()
  {
    if (Size <= 0) { throw std::logic_error ("[Size is =< 0]"); }

    for (size_t i = 0; i < Size; ++i)
    {
      FormulaArray[i].Apply ();
    }
  }
}

File: P5/Plan.cs
/// <file> Plan.cs </file>
/// <brief>
/// [DESC]: This file contains the definition of the Plan class, which encapsulates an array of Formula
///        Objects. The 'Plan' class brings more functionality than a C-style array with methods like
///        'AddFormula(...)', 'RemoveLastFormula()', and so on. Memory management is abstracted and
///        hidden from the client. The class supports shallow and efficient deep copying with move
///        semantics. <see cref="[MOVE SEMANTICS]"/> The class also supports some simple mathematical 
///        operations through overloaded operators. <see cref="[OPERATORS]"/>
/// </brief>
/// 
/// <author>Jakob Balkovec (CPSC 3200)</author>
/// <instructor>A. Dingle (CPSC 3200)</instructor>
/// <date> Fri 14th Nov </date>
/// 
/// <version>
///     Revision History
///     - 1.0 [27/10/2023]: Optimization and Improved Move Semantics
///     - 2.0 [28/10/2023]: Debugging
///     - 3.0 [28/10/2023]: Documentation
///     - 4.0 [14/10/2023]: C++ -> C# (operators)
/// </version>
/// 
/// <invariant>Capacity is the capacity for FormulaArray and should be greater than or equal to 2.</invariant>
/// <invariant>Size of Plan and should be greater than or equal to 1.</invariant>
/// <invariant>FormulaArray is a dynamic array to store Formula objects.</invariant>
/// <invariant>FormulaArray cannot be nullptr</invariant>
/// 
/// <moveSemantics>
///     Plan Obj1 = new Plan();
///     Plan Obj2 = Obj1.PShallowCopy();
///     Plan Obj3 = Obj1.PDeepCopy();
/// </moveSemantics>
/// 
/// <dependencies>
///     <internal>
///         - Member, Helper, Utility methods
///         - Private Data Members
///     </internal>
///     <external>
///         - 'Formula' class
///     </external>
/// </dependencies>
/// 
/// <namespace>
///     Might be deemed unnecessary, but despite the distinctive function names, global
///     namespace pollution is still in the picture.
///     - {ResourceConversion} Encapsulates the 'Formula', 'Plan', 'ExecutablePlan' class.
/// </namespace>

using System;

namespace ResourceConversion
{
    public class Plan : IPlan
    {

        ///! Member
        protected Formula[] FormulaArray { get; private set; }

        /// <summary>
        /// - If set to true, it produces console output
        /// </summary>
        private readonly bool DisplayValues = true;
        //! Logs


        /// <summary>
        /// - Initializes a new instance of the Plan class with an array of formulas.
        /// </summary>
        /// 
        /// <param name="FormulaArray_">
        /// - The array of formulas.
        /// - The FormulaArray_ parameter is an array of Formula objects, each representing a formula
        ///   used within the Plan.
        /// </param>
        /// 
        /// <exception cref="ArgumentException">
        /// - Thrown if the provided array contains null values.
        /// </exception>
        /// 
        /// <pre>Pre-Condition:
        /// - The FormulaArray_ parameter must be instantiated before calling this constructor.
        /// - FormulaArray_ must not include any null values.
        /// </pre>
        /// 
        /// <post>Post-Condition:
        /// - An instance of the Plan object is created with the specified array of formulas.
        /// </post>
        /// 
        /// <remarks>
        /// - This constructor initializes a new instance of the Plan class by accepting an array of formulas
        ///   as a parameter. It ensures that the array is free of null values and throws an ArgumentException
        ///   if any null values are found. After validation, the constructor initializes the Plan object
        ///   and logs the construction if LogMessages is enabled.
        /// </remarks>
        public Plan(Formula[] FormulaArray_)
        {
            if (FormulaArray_.All(resource => resource == null))
            {
                throw new ArgumentException("Raise: => Plan() [Array cannot contain null values]");
            }

            this.FormulaArray = FormulaArray_;
        }

        /// <summary>
        /// - Creates a deep copy of the Plan instance, including deep copies of its formulas.
        /// </summary>
        /// 
        /// <returns>
        /// - A deep copy of the Plan instance.
        /// </returns>
        /// 
        /// <remarks>
        /// - The Clone method creates a deep copy of the current Plan instance, including creating
        ///   deep copies of all its associated Formula objects. This means that all Formula objects
        ///   within the new Plan instance are independent copies of the original, preserving the state
        ///   of each formula without any reference to the original Plan instance.
        /// </remarks>
        /// 
        /// <pre>Pre-Condition:
        /// - The Plan instance must be properly initialized before invoking this method.
        /// </pre>
        /// 
        /// <post>Post-Condition:
        /// - A new Plan instance is created as a deep copy of the current instance and its formulas.
        /// </post>
        ///
        /// <attention> Client has to cast the object to the appropriaet type </attention>
        private object Clone()
        {
            Formula[] ClonedFormulas = new Formula[FormulaArray.Length];
            for (int i = 0; i < FormulaArray.Length; i++)
            {
                ClonedFormulas[i] = FormulaArray[i].Clone();
            }
            return new Plan(ClonedFormulas);
        }

        /// <summary>
        /// - Creates a deep copy of the Plan instance, including deep copies of its formulas.
        /// </summary>
        /// 
        /// <returns>
        /// - A deep copy of the Plan instance.
        /// </returns>
        /// 
        /// <remarks>
        /// - The PlanDeepCopy method is used to create a deep copy of the current Plan instance,
        ///   including deep copies of all its associated Formula objects. The deep copy ensures that
        ///   the new Plan instance is entirely independent and self-contained, preserving the state of
        ///   each formula without any reference to the original Plan instance.
        /// </remarks>
        /// 
        /// <pre>Pre-Condition:
        /// - The current Plan instance (this) cannot be null.
        /// - The Clone method must be available for invocation to create deep copies of the formulas.
        /// </pre>
        /// 
        /// <post>Post-Condition:
        /// - A deep copy of the Plan instance, along with its associated formulas, is successfully created.
        /// </post>
        ///
        /// <attention>
        /// - Declared Nullable
        /// - Client has to cast the object to the appropriaet type
        /// </attention>
        public virtual object? DeepCopy() { return this.Clone() as Plan; }

        /// <summary>
        /// - Creates a shallow copy of the Plan instance. The formulas are not deep-copied.
        /// </summary>
        /// 
        /// <returns>
        /// - A shallow copy of the Plan instance.
        /// </returns>
        /// 
        /// <remarks>
        /// - The PlanShallowCopy method is used to create a shallow copy of the current Plan instance.
        /// - Unlike a deep copy, the formulas within the Plan are not deep-copied, and both the original
        ///   and copied Plan instances will share references to the same Formula objects.
        /// - This means changes to the formulas in the copied Plan will affect the original and vice versa.
        /// </remarks>
        /// 
        /// <pre>Pre-Condition:
        /// - The current Plan instance (this) cannot be null.
        /// - The `MemberwiseClone` method must be available for invocation to create shallow copies.
        /// </pre>
        /// 
        /// <post>Post-Condition:
        /// - A shallow copy of the Plan instance is successfully created.
        /// </post>
        ///
        /// <attention>
        /// - Declared Nullable
        /// - Client has to cast the object to the appropriaet type
        /// </attention>
        public virtual object? ShallowCopy() { return this.MemberwiseClone() as Plan; }

        /// <summary>
        /// - Determines whether the current Plan instance is equal to another Plan instance.
        /// </summary>
        /// 
        /// <param name="obj">
        /// - The Plan instance to compare with the current instance.
        /// </param>
        /// 
        /// <returns>
        /// <c>true</c>
        /// - if the two instances are considered equal;
        /// otherwise,
        /// <c>false</c>.
        /// </returns>
        /// 
        /// <remarks>
        /// Two Plan instances are considered equal if they have the same content in their FormulaArray property.
        /// This method overrides the <see cref="object.Equals(object)"/> method.
        /// </remarks>
        public new bool Equals(object? obj)
        {
            if (obj is Plan otherPlan)
            {
                if (ReferenceEquals(this, otherPlan)) { return true; }

                if (FormulaArray.Length != otherPlan.FormulaArray.Length) { return false; }

                for (int i = 0; i < FormulaArray.Length; i++)
                {
                    if (!FormulaArray[i].Equals(otherPlan.FormulaArray[i])) { return false; }
                }
                return true;
            }
            return false;
        }

        /// <summary>
        /// - Adds a new formula to the plan.
        /// </summary>
        /// 
        /// <param name="NewFormula">
        /// - The formula to add.
        /// </param>
        /// 
        /// <exception cref="ArgumentNullException">
        /// - Thrown if NewFormula is null.
        /// </exception>
        /// 
        /// <remarks>
        /// - The Method method is used to add a new formula to the Plan. This method allows you to expand
        ///   the Plan's list of formulas. The 'in' keyword specifies that NewFormula is passed by reference
        ///   and treated as a constant, meaning it cannot be modified within the method.
        /// </remarks>
        /// 
        /// <seealso cref="https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/in"/>
        /// <seealso cref="https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/ranges"/>
        /// 
        /// <pre>Pre-condition:
        /// - NewFormula must not be null.
        /// </pre>
        /// 
        /// <post>Post-condition:
        /// - The array of formulas will contain one additional formula.
        /// - The NewFormula will be added to the end of the FormulaArray.
        /// </post>
        /// 
        /// </remarks>
        public virtual void AddFormula(in Formula NewFormula)
        {
            Formula[] NewFormulaArray = new Formula[FormulaArray.Length + 1];

            for (uint i = 0; i < FormulaArray.Length; i++)
            {
                NewFormulaArray[i] = FormulaArray[i];
            }

            NewFormulaArray[^1] = NewFormula ?? throw new ArgumentNullException("Raise: AddFormula() => [NewFormula is 'null']");
            FormulaArray = NewFormulaArray;
        }

        /// <summary>
        /// - Removes the last formula from the plan.
        /// </summary>
        /// 
        /// <exception cref="InvalidDataException">
        /// - Thrown if the array size is less than or equal to 0.
        /// </exception>
        /// 
        /// <pre>Pre-condition:
        /// - The array of formulas must not be empty.
        /// </pre>
        /// 
        /// <post>Post-condition:
        /// - The array of formulas will contain one less formula.
        /// - The last formula in the array will be removed.
        /// </post>
        /// 
        /// <remarks>
        /// - The Method method is used to remove the last formula from the Plan. This method allows you to
        ///   reduce the Plan's list of formulas by eliminating the last formula in the array.
        /// </remarks>
        public virtual void RemoveLastFormula()
        {
            if (FormulaArray.Length <= 0)
            {
                throw new InvalidDataException("Raise: RemoveLastFormula() => [Array size is less than or equal to 0]");
            }

            Formula[] NewFormulaArray = new Formula[FormulaArray.Length - 1];
            for (uint i = 0; i < NewFormulaArray.Length; i++)
            {
                NewFormulaArray[i] = FormulaArray[i];
            }

            FormulaArray = NewFormulaArray;
        }

        /// <summary>
        /// - Replaces a formula at the specified index with a new formula.
        /// </summary>
        /// 
        /// <param name="Index">
        /// - The index of the formula to replace.
        /// </param>
        /// 
        /// <param name="NewFormula">
        /// - The new formula to replace with.
        /// </param>
        /// 
        /// <exception cref="IndexOutOfRangeException">
        /// - Thrown if Index is out of bounds.
        /// </exception>
        /// 
        /// <exception cref="ArgumentNullException">
        /// - Thrown if NewFormula is 'null'.
        /// </exception>
        /// 
        /// <pre>Pre-condition:
        /// - Index must be a valid index within the array of formulas.
        /// - NewFormula must be passed in as a valid parameter and not be 'null'.
        /// </pre>
        /// 
        /// <post>Post-condition:
        /// - The formula at the specified Index will be replaced with NewFormula.
        /// </post>
        /// 
        /// <remarks>
        /// - The ReplaceFormula allows the client to replace a formula at the specified index within the Plan's
        ///   list of formulas with a new formula. This method ensures that the index is within bounds and
        ///   that the NewFormula parameter is not null before performing the replacement.
        /// </remarks>
        public virtual void ReplaceFormula(Formula NewFormula, ushort Index)
        {

            if (Index < 0 || Index >= FormulaArray.Length)
            {
                throw new IndexOutOfRangeException("Raise: ReplaceFormula() => [Index is out of bounds]");
            }

            if (NewFormula is null)
            {
                throw new ArgumentNullException("Raise: ReplaceFormula() => [NewFormula cannot be 'null']");
            }

            FormulaArray[Index] = NewFormula;
        }

        /// <summary>
        /// - Applies each formula within the Plan, triggering the calculation or processing
        ///   of each formula's output.
        /// </summary>
        /// 
        /// <remarks>
        /// - The PlanApply method iterates through each formula in the Plan's array and invokes
        ///   the Apply method of each formula. This method is typically used to calculate or
        ///   process the output of each formula within the Plan.
        /// </remarks>
        /// 
        /// <pre>Pre-condition:
        /// - The Plan instance must be properly initialized before calling this method.
        /// </pre>
        /// 
        /// <post>Post-condition:
        /// - Each formula's Apply method has been executed, potentially updating the formula's
        /// output.
        /// </post>
        public virtual void Apply()
        {
            for (int i = 0; i < FormulaArray.Length; i++) { FormulaArray[i].Apply(); }
        }

        /// <summary>
        ///  - Displays the resources used by each formula within the Plan, if DisplayValues is enabled.
        /// </summary>
        /// 
        /// <remarks>
        /// - The PlanDisplayResources method is responsible for displaying the resources used by
        ///   each formula in the Plan's array. This information is displayed only if the DisplayValues
        ///   option is enabled.
        /// </remarks>
        /// 
        /// <pre>Pre-condition:
        /// - The Plan instance must be properly initialized before calling this method.
        /// </pre>
        /// 
        /// <post>Post-condition:
        /// - If DisplayValues is enabled, the resources used by each formula are displayed.
        /// </post>
        public void PlanDisplayResources()
        {
            if (DisplayValues)
            {
                for (uint i = 0; i < FormulaArray.Length; i++)
                {
                    FormulaArray[i].DisplayResources();
                }
            }
            return;
        }

        /// <summary>
        /// - Displays the output of each formula within the Plan, if DisplayValues is enabled.
        /// </summary>
        /// 
        /// <remarks>
        /// - The PlanDisplayFormulaOutput method is responsible for displaying the output of each
        ///   formula in the Plan's array. This output is displayed only if the DisplayValues option is
        ///   enabled.
        /// </remarks>
        /// 
        /// <pre>Pre-condition:
        /// - The Plan instance must be properly initialized before calling this method.
        /// </pre>
        /// 
        /// <post>Post-condition:
        /// - If DisplayValues is enabled, the output of each formula is displayed.
        /// </post>
        public void PlanDisplayFormulaOutput()
        {
            if (DisplayValues)
            {
                for (uint i = 0; i < FormulaArray.Length; i++)
                {
                    FormulaArray[i].DisplayFormulaOutput();
                }
            }
            return;
        }

        /// <summary>
        /// - Gets an array of Formula objects.
        /// </summary>
        /// 
        /// <returns>
        /// - An array of Formula objects.
        /// </returns>
        public Formula[] GetFormulaArray() { return FormulaArray; }
    }
}


